### ISSUE TITLE
Proposal: refManualReset

### ISSUE THREAD

**Issue by pabueco** [vueuse/vueuse#1479]
### Clear and concise description of the problem

As a developer using VueUse I might want a `refManualReset` function so that I can initialize a ref with a value and reset it later to that initial value.

While building some browser games I constantly had the case that I needed some initial values for variables and had to reset **some** of them later to their initial value. So simply moving the initialization of these variables into a function wasn't really a nice option. Instead I created a really simple function that uses `extendRef` to add the `reset()` method. It works really well for the use cases I have so I wanted to see if this would be helpful to others as well.

#### Usage
```typescript
const myVariable = refManualReset(5)
myVariable.value = 10
myVariable.reset() // value is 5 again
```

### Suggested solution

A really simple implementation could look like this:

```typescript
import { extendRef } from '@vueuse/shared'
import { ref, UnwrapRef } from 'vue'

export function refManualReset<T>(defaultValue?: T): Ref<T> {
  const valueRef = ref<T>(defaultValue)

  const reset = () => {
    valueRef.value = defaultValue as UnwrapRef<T>
  }

  const valueExtendedRef = extendRef(valueRef, {
    reset,
  })

  return valueExtendedRef
}
```

Obviously that would create issues when using it for anything else than primitive values. So to support objects and arrays some kind of cloning would be needed. This could be implemented like in `useManualRefHistory`.

If this idea is something that would make sense to be added I would be happy to create a PR with a better implementation.

### Alternative

You can of course just use a different variable/constant that holds the value and use that to manually change the variable when it needs to be reset, but I found it much cleaner and clearer to just call `myVariable.reset()` instead.

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vueuse/vueuse/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vueuse/vueuse/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vueuse.org/guide).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

**Comment by okxiaoliang4** [comment]
Can `useManualRefHistory` meet your needs.

```ts
import { ref } from 'vue'
import { useManualRefHistory } from '@vueuse/core'

const myVariable = ref(0)
const { history, commit, undo, redo } = useManualRefHistory(myVariable)
commit()
myVariable.value = 10

undo()
// myVariable 0
```

Or maybe we could add an option `immediate`, like this:

```ts
import { ref } from 'vue'
import { useManualRefHistory } from '@vueuse/core'

const myVariable = ref(0)
const { history, commit, undo, redo } = useManualRefHistory(myVariable, { immediate: true })
myVariable.value = 10

undo()
// myVariable 0
```

**Comment by pabueco** [comment]
@okxiaoliang4 Thanks, I considered using `useManualRefHistory` but I found the approach a bit unclear and hard to follow when looking at the code because the intention is not really clearly visible. But maybe that's just me and it does not make sense to add a separate function for this.
The `immediate` option would be a good first step. Another idea would be to add some kind of `reset` function to `useManualRefHistory` to undo all the changes and go back to the initial value/first commit.

**Comment by okxiaoliang4** [comment]
> @okxiaoliang4 Thanks, I considered using `useManualRefHistory` but I found the approach in my case a bit unclear and hard to follow when looking at the code because the intention is not really clearly visible. But maybe that's just me and it does not make sense to add a separate function to the core for this.
> 
> The `immediate` option would be a good first step. Another idea would be to add some kind of `reset` function to `useManualRefHistory` to undo all the changes and go back to the initial value/first commit.

or this?
```ts
import { ref } from 'vue'
import { useManualRefHistory } from '@vueuse/core'

const myVariable = ref(0)
const { undo: rest } = useManualRefHistory(myVariable, { immediate: true })
myVariable.value = 10

rest()
// myVariable 0
```

**Comment by pabueco** [comment]
That's an interesting idea and would be probably good enough in most cases. I still think it would be cleaner to provide a separate function to set the history back to the initial value, so it would work also together with using commits.

For me my existing solution works fine, so I'm probably gonna continue using it, but I understand if that's not something that makes sense to be added to the package.

**Comment by WingDust** [comment]
@okxiaoliang4 I think  ` add some kind of reset function to useManualRefHistory to undo all the changes and go back to the initial value/first commi` is better, because Resetting should be the default action of historical control

**Comment by antfu** [comment]
I am fine to have it if we address the following issues:

- The downside of using `extendRef` is that you can't call the reset function in the template
- Should we have an option to update the initial value?

**Comment by WingDust** [comment]
@antfu 
packages/core/useManualRefHistory/index.ts
```ts
import { isFunction, timestamp } from '@vueuse/shared'
import type { Ref } from 'vue-demi'
import { computed, markRaw, ref } from 'vue-demi'

export interface UseRefHistoryRecord<T> {
  snapshot: T
  timestamp: number
}

export type CloneFn<F, T = F> = (x: F) => T

export interface UseManualRefHistoryOptions<Raw, Serialized = Raw> {
  /**
   * Maximum number of history to be kept. Default to unlimited.
   */
  capacity?: number

  /**
   * Clone when taking a snapshot, shortcut for dump: JSON.parse(JSON.stringify(value)).
   * Default to false
   *
   * @default false
   */
  clone?: boolean | CloneFn<Raw>
  /**
   * Serialize data into the history
   */
  dump?: (v: Raw) => Serialized
  /**
   * Deserialize data from the history
   */
  parse?: (v: Serialized) => Raw

  /**
   * Deserialize data from the history
   */
  setSource?: (source: Ref<Raw>, v: Raw) => void
}

export interface UseManualRefHistoryReturn<Raw, Serialized> {
  /**
   * Bypassed tracking ref from the argument
   */
  source: Ref<Raw>

  /**
   * An array of history records for undo, newest comes to first
   */
  history: Ref<UseRefHistoryRecord<Serialized>[]>

  /**
  * Last history point, source can be different if paused
  */
  last: Ref<UseRefHistoryRecord<Serialized>>

  /**
   * Same as 'history'
   */
  undoStack: Ref<UseRefHistoryRecord<Serialized>[]>

  /**
   * Records array for redo
   */
  redoStack: Ref<UseRefHistoryRecord<Serialized>[]>

  /**
   * A ref representing if undo is possible (non empty undoStack)
   */
  canUndo: Ref<boolean>

  /**
   * A ref representing if redo is possible (non empty redoStack)
   */
  canRedo: Ref<boolean>

  /**
   * Undo changes
   */
  undo(): void

  /**
   * Redo changes
   */
  redo(): void

  /**
   * Clear all the history
   */
  clear(): void

  /**
   * Create new a new history record
   */
  commit(): void

  /**
   * Reset ref's value with lastest history
   */
  reset(): void
  /**
   * re set ref value to default
   */
  reDefault(): UseRefHistoryRecord<Serialized>[]
}

const fnClone = <F, T>(v: F): T => JSON.parse(JSON.stringify(v))
const fnBypass = <F, T>(v: F) => v as unknown as T
const fnSetSource = <F>(source: Ref<F>, value: F) => source.value = value

type FnCloneOrBypass<F, T> = (v: F) => T

function defaultDump<R, S>(clone?: boolean | CloneFn<R>) {
  return (clone ? isFunction(clone) ? clone : fnClone : fnBypass) as unknown as FnCloneOrBypass<R, S>
}

function defaultParse<R, S>(clone?: boolean | CloneFn<R>) {
  return (clone ? isFunction(clone) ? clone : fnClone : fnBypass) as unknown as FnCloneOrBypass<S, R>
}

/**
 * Track the change history of a ref, also provides undo and redo functionality.
 *
 * @see https://vueuse.org/useManualRefHistory
 * @param source
 * @param options
 */
export function useManualRefHistory<Raw, Serialized = Raw>(
  source: Ref<Raw>,
  options: UseManualRefHistoryOptions<Raw, Serialized> = {},
): UseManualRefHistoryReturn<Raw, Serialized> {
  const {
    clone = false,
    dump = defaultDump<Raw, Serialized>(clone),
    parse = defaultParse<Raw, Serialized>(clone),
    setSource = fnSetSource,
  } = options

  function _createHistoryRecord(): UseRefHistoryRecord<Serialized> {
    return markRaw({
      snapshot: dump(source.value),
      timestamp: timestamp(),
    })
  }

  const last: Ref<UseRefHistoryRecord<Serialized>> = ref(_createHistoryRecord()) as Ref<UseRefHistoryRecord<Serialized>>

  const undoStack: Ref<UseRefHistoryRecord<Serialized>[]> = ref([])
  const redoStack: Ref<UseRefHistoryRecord<Serialized>[]> = ref([])

  const _setSource = (record: UseRefHistoryRecord<Serialized>) => {
    setSource(source, parse(record.snapshot))
    last.value = record
  }

  const commit = () => {
    undoStack.value.unshift(last.value)
    last.value = _createHistoryRecord()

    if (options.capacity && undoStack.value.length > options.capacity)
      undoStack.value.splice(options.capacity, Infinity)
    if (redoStack.value.length)
      redoStack.value.splice(0, redoStack.value.length)
  }

  const clear = () => {
    undoStack.value.splice(0, undoStack.value.length)
    redoStack.value.splice(0, redoStack.value.length)
  }

  const undo = () => {
    const state = undoStack.value.shift()

    if (state) {
      redoStack.value.unshift(last.value)
      _setSource(state)
    }
  }

  const redo = () => {
    const state = redoStack.value.shift()

    if (state) {
      undoStack.value.unshift(last.value)
      _setSource(state)
    }
  }

  const reset = () => {
    _setSource(last.value)
  }
  const reDefault = ()=>{
     const lastHistory =undoStack.value.map(i=>i)
     const state = undoStack.value.pop()

     if (state){
      _setSource(state)
      clear()
     }
     return lastHistory
  }

  const history = computed(() => [last.value, ...undoStack.value])

  const canUndo = computed(() => undoStack.value.length > 0)
  const canRedo = computed(() => redoStack.value.length > 0)

  return {
    source,
    undoStack,
    redoStack,
    last,
    history,
    canUndo,
    canRedo,

    clear,
    commit,
    reset,
    undo,
    redo,
    reDefault
  }
}

``` 
unit test
```ts
  test('re set ref value to default',()=>{
    const v = ref(0)
    const { history, commit, undoStack, redoStack, reset, undo,reDefault } = useManualRefHistory(v)
    v.value = 1
    commit()
    v.value = 2
    commit()
    v.value = 3
    commit()
    v.value = 4
    commit()
    reDefault()
    console.log(history);
    expect(v.value).toBe(0)
    expect(history.value[0].snapshot).toBe(0)
  })
```
i add a `reDefault` method to reset to default value and return last history
it 's ok?
